core/access/peer.go

```diff
@@ -0,0 +1,181 @@
+	GetProofs      getProofsFn
+
+	version int // LES protocol version number
+}
+
+// newPeer creates a new ODR peer
+func newPeer(id string, version int, head common.Hash, getBlockBodies getBlockBodiesFn, getNodeData getNodeDataFn, getReceipts getReceiptsFn, getProofs getProofsFn) *Peer {
+	return &Peer{
+		id:             id,
+		head:           head,
+		GetBlockBodies: getBlockBodies,
+		GetNodeData:    getNodeData,
+		GetReceipts:    getReceipts,
+		GetProofs:      getProofs,
+		version:        version,
+	}
+}
+
+// Id returns a peer's ID string
+func (p *Peer) Id() string {
+	return p.id
+}
+
+
+// Promote increases the peer's reputation
+func (p *Peer) Promote() {
+	atomic.AddInt32(&p.rep, 1)
+}
+
+// Demote decreases the peer's reputation or leaves it at 0.
+func (p *Peer) Demote() {
+	for {
+		// Calculate the new reputation value
+		prev := atomic.LoadInt32(&p.rep)
+		next := prev / 2
+
+		// Try to update the old value
+		if atomic.CompareAndSwapInt32(&p.rep, prev, next) {
+			return
+		}
+	}
+}
+
+// peerSet represents the collection of active peer participating in the block
+// download procedure.
+type peerSet struct {
+	peers map[string]*Peer
+	lock  sync.RWMutex
+}
+
+// newPeerSet creates a new peer set top track the active download sources.
+func newPeerSet() *peerSet {
+	return &peerSet{
+		peers: make(map[string]*Peer),
+	}
+}
+
+// Register injects a new peer into the working set, or returns an error if the
+// peer is already known.
+func (ps *peerSet) Register(p *Peer) error {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+
+	if _, ok := ps.peers[p.id]; ok {
+		return errAlreadyRegistered
+	}
+	ps.peers[p.id] = p
+	return nil
+}
+
+// Unregister removes a remote peer from the active set, disabling any further
+// actions to/from that particular entity.
+func (ps *peerSet) Unregister(id string) error {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+
+	if _, ok := ps.peers[id]; !ok {
+		return errNotRegistered
+	}
+	delete(ps.peers, id)
+	return nil
+}
+
+// Peer retrieves the registered peer with the given id.
+func (ps *peerSet) Peer(id string) *Peer {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	return ps.peers[id]
+}
+
+// Len returns if the current number of peers in the set.
+func (ps *peerSet) Len() int {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	return len(ps.peers)
+}
+
+// AllPeers retrieves a flat list of all the peers within the set.
+func (ps *peerSet) AllPeers() []*Peer {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	list := make([]*Peer, 0, len(ps.peers))
+	for _, p := range ps.peers {
+		list = append(list, p)
+	}
+	return list
+}
+
+// BestPeers returns an ordered list of available peers, starting with the
+// highest reputation
+func (ps *peerSet) BestPeers() []*Peer {
+	list := ps.AllPeers()
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	for i := 0; i < len(list); i++ {
+		for j := i + 1; j < len(list); j++ {
+			if atomic.LoadInt32(&list[i].rep) < atomic.LoadInt32(&list[j].rep) {
```

Comment:
This sort is unsound. I suggest storing reputation in a `map[id]int` on peerSet and  protect the sort with the peerSet mutex. `promote(peer)` and `demote(peer)` should be implemented on `peerSet`.

Also consider using `sort` with:

```go
func (ps *peerSet) bestN(n int) []peer {
	ps.lock.RLock()
	defer ps.lock.RUnlock()
	list := make([]peer, len(ps.peers))
	for i, p := range ps.peers {
		list[i] = p
	}
	s = &byRep{list, self}
	sort.Sort(s)
	return s.data[:n]
}

// sort interface functions
// caller holds the lock

type byRep struct {
	list
	*peerSet
}

func (ps *byRep) Len() {
	return len(ps.list)
}

func (ps *byRep) Less(i, j int) bool {
	return ps.rep[ps.list[i].id()] < ps.rep[ps.list[j].id()]
}

func (ps *byRep) Swap(i, j int) {
	ps.list[i] = ps.list[j]
}

```